File: CalendarWithObjC/CalendarWithObjC/CustomCalendar/CalendarViewController.m
//
//  CalendarViewController.m
//  CathayBank
//
//  Created by Marvin iOS on 2018/9/12.
//  Copyright © 2018年 YuChen Hsu. All rights reserved.
//

#import "CalendarViewController.h"


@interface CalendarViewController ()

@end

@implementation CalendarViewController

@synthesize delegate, calendarType, firstDate, customDuration;

- (void)viewDidLoad {
    [super viewDidLoad];
    monthsString = [NSArray new];
    calendarModels = [NSMutableArray new];
    selectedIndexes = [NSMutableArray new];
    today = [NSDate date];
    
    if (customDuration && customDuration.count > 0) {
        // 有選擇過日期
        if (customDuration.count == 2) {
            // 起迄都選過了,依照進入點
            
            NSDate *startDate = customDuration[@"start"];
            NSDate *endDate = customDuration[@"end"];
            
            if (calendarType == Start && startDate) {
                firstDate = startDate;
            }
            else {
                firstDate = endDate;
            }
        }
        else {
            //只選過一個
            
            NSArray *keys = customDuration.allKeys;
            firstDate = customDuration[keys[0]];
        }
        
    }
    else {
        // 如果沒有選擇,就從今天
        firstDate = [NSDate date];
    }
    
    [self registerCell];
    [self setupUI];
    
    calendarProvider = [CalendarProvider new];
    
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    [self.calendarCollectionView reloadData];
    [self initCalendar];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    
    // 防止跑版
    if (@available(iOS 11.0, *)) {
        _containerTopConstraint.constant = 57.0;
    }
    else {
        _containerTopConstraint.constant = 77.0;
    }
    
    [self.view layoutIfNeeded];
}

#pragma mark - Setup UI
- (void)registerCell {
    
    cellIdentifier = @"DateCell";
    [_calendarCollectionView registerNib:[UINib nibWithNibName:@"CalendarCollectionViewCell" bundle:nil] forCellWithReuseIdentifier:cellIdentifier];
    
    headerIdentifier = @"HeaderCell";
    [_calendarCollectionView registerNib:[UINib nibWithNibName:@"CalendarHeaderCollectionViewCell" bundle:nil] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:headerIdentifier];
}

- (void)setupUI {
    [_cancelButton setTitle:@"Cancel" forState:UIControlStateNormal];
    [_okButton setTitle:@"Confirm" forState:UIControlStateNormal];
    [_sundayLabel setText:@"Sun"];
    [_mondayLabel setText:@"Mon"];
    [_tuesdayLabel setText:@"Tue"];
    [_wednesdayLabel setText:@"Wed"];
    [_thursdayLabel setText:@"Thu"];
    [_fridayLabel setText:@"Fri"];
    [_saturdayLabel setText:@"Sat"];
    
    _cancelButton.layer.borderWidth = 1.0;
    _cancelButton.layer.borderColor = [UIColor colorWithRed:0.0f/255.0f green:158.0f/255.0f blue:73.0f/255.0f alpha:1.0f].CGColor;
    
    shadowView.layer.shadowRadius  = 5;
    shadowView.layer.shadowColor   = [UIColor blackColor].CGColor;
    shadowView.layer.shadowOffset  = CGSizeMake(0.0f, 0.0f);
    shadowView.layer.shadowOpacity = 0.1f;
    shadowView.layer.masksToBounds = NO;
    
    UIEdgeInsets shadowInsets     = UIEdgeInsetsMake(0, 0, -1.5f, 0);
    UIBezierPath *shadowPath      = [UIBezierPath bezierPathWithRect:UIEdgeInsetsInsetRect(shadowView.bounds, shadowInsets)];
    shadowView.layer.shadowPath    = shadowPath.CGPath;
    
}


#pragma mark - Override method
- (NSInteger)maxYear {
    return NSIntegerMax;
}

- (NSInteger)minYear {
    return 1975;
}

- (CalendarCollectionViewCell *)canbeSelectCell:(CalendarCollectionViewCell *) cell
                                     monthModel:(CalendarMonthModel *) model
                                            day:(NSInteger) day {
    //child class canbe override method
    //implement on child class
    [cell setUserInteractionEnabled:YES];
    [cell setNormalState];
    return cell;
}

//- (CalendarCollectionViewCell *)canbeSelectCell:(CalendarCollectionViewCell *)cell
//                                     monthModel:(CalendarMonthModel *)model
//                                            day:(NSInteger)day {
//    NSDate *dateAtCell = [DateGenerator getDateFromInt:model.year month:model.month day:day];
//
//    // 這邊處理起迄不同要反灰的 UI
//
//    NSDate *compareDate;
//    NSDate *threeMonthLimit;
//    NSDate *oneYearLimit = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitMonth value:-12 toDate:today options:0];
//
//    if (self.calendarType == Start) {
//
//        if (self.customDuration[@"end"]) {
//            compareDate = self.customDuration[@"end"];
//            threeMonthLimit = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitMonth value:-3 toDate:compareDate options:0];
//        }
//        else
//        {
//            compareDate = self.firstDate;
//        }
//
//        //        if ([dateAtCell compare:compareDate] == NSOrderedDescending && self.customDuration.count > 0) {
//        //            // 在起始日之前,灰色且不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else if ([dateAtCell compare:today] == NSOrderedDescending) {
//        //            // 在今天之後的日期不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else if (threeMonthLimit && [dateAtCell compare:threeMonthLimit] == NSOrderedAscending) {
//        //            // 超過迄日前三個月,灰色且不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else if (oneYearLimit && [dateAtCell compare:oneYearLimit] == NSOrderedAscending) {
//        //            // 超過當日前一年,灰色且不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else {
//        [cell setUserInteractionEnabled:YES];
//        [cell setNormalState];
//        //        }
//    }
//    else {
//
//        if (self.customDuration[@"start"]) {
//            compareDate = self.customDuration[@"start"];
//            threeMonthLimit = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitMonth value:3 toDate:compareDate options:0];
//        }
//        else {
//            compareDate = self.firstDate;
//        }
//
//        //        if ([dateAtCell compare:compareDate] == NSOrderedAscending && self.customDuration.count > 0) {
//        //            // 在迄日之後,灰色且不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else if (threeMonthLimit && [dateAtCell compare:threeMonthLimit] == NSOrderedDescending) {
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else if (oneYearLimit && [dateAtCell compare:oneYearLimit] == NSOrderedAscending) {
//        //            // 超過當日前一年,灰色且不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else if ([dateAtCell compare:today] == NSOrderedDescending) {
//        //            // 在今天之後的日期不能選
//        //            [cell setUserInteractionEnabled:NO];
//        //            [cell setUnableState];
//        //        }
//        //        else {
//        [cell setUserInteractionEnabled:YES];
//        [cell setNormalState];
//        //        }
//    }
//
//    return cell;
//}

#pragma mark - UICollectionViewFlowlayoutDelegate
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    return CGSizeMake(_calendarCollectionView.frame.size.width, 50.0);
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return 0.0f;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return 0.0f;
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat itemWidth = _calendarCollectionView.frame.size.width / 7.0;
    
    CGFloat itemHeight = itemWidth;
    
    return CGSizeMake(itemWidth, itemHeight);
}

#pragma mark - CollectionView Delegate
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    CalendarCollectionViewCell *cell = (CalendarCollectionViewCell*)[collectionView cellForItemAtIndexPath:indexPath];
    
    if ([selectedIndexes containsObject:indexPath]&&selectedIndexes.count!=0) {
        [cell setNormalState];
        [selectedIndexes removeObject:indexPath];
        selectedDate = nil;
    }
    else {
        // 只讓使用者選 1 點
        if (selectedIndexes.count > 0) {
            NSIndexPath *previousIndex = [selectedIndexes firstObject];
            
            if ([_calendarCollectionView.indexPathsForVisibleItems containsObject:previousIndex]) {
                CalendarCollectionViewCell *previousCell = (CalendarCollectionViewCell*)[collectionView cellForItemAtIndexPath:previousIndex];
                [previousCell setNormalState];
            }
            
            [selectedIndexes removeAllObjects];
            [selectedIndexes addObject:indexPath];
            [cell setSelectedState];
        }
        else {
            [selectedIndexes addObject:indexPath];
            [cell setSelectedState];
        }
    }
    
    CalendarMonthModel *monthModel = calendarModels[indexPath.section];
    
    NSInteger dayInt = [cell.dateLabel.text integerValue];
    
    if (monthModel && dayInt) {
        selectedDate = [DateGenerator getDateFromInt:monthModel.year month:monthModel.month day:dayInt];
    }
        
        // 下面是要選 2 點的邏輯,未來如果確定了就打開
//        if (selectedIndexes.count > 2) {
//            NSArray *sortedArray = [selectedIndexes sortedArrayUsingSelector:@selector(compare:)];
//            selectedIndexes = [NSMutableArray arrayWithArray:sortedArray];
//            NSIndexPath *lastIndexPath = [selectedIndexes lastObject];
//
//            NSIndexPath *deselectIndexPath;
        
            // 有可能 UI 邏輯還會更改,之後再修
//            if (indexPath == lastIndexPath) {
//                // 最後一個點擊 index 在前兩點之後,則移掉中間那個 index 為 1 的
//                deselectIndexPath = [sortedArray objectAtIndex:1];
//            }
//            else if (indexPath == [selectedIndexes firstObject]) {
//                // 最後一個點擊的在 index 之前,則移掉最後面那個
//                deselectIndexPath = lastIndexPath;
//            }
//            else {
//                // 最後一個點擊的,在兩者之間,則移掉最後一個
//                deselectIndexPath = lastIndexPath;
//            }
//
//            [selectedIndexes removeObject:deselectIndexPath];
//
//            if ([_calendarCollectionView.indexPathsForVisibleItems containsObject:deselectIndexPath]) {
//                CalendarCollectionViewCell *deselectCell = (CalendarCollectionViewCell*)[collectionView cellForItemAtIndexPath:deselectIndexPath];
//                [deselectCell setNormalState];
//            }1
//    }
}

#pragma mark - CollectionView DataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return calendarModels.count;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    
    if (!calendarModels || calendarModels.count == 0) {
        return 0;
    }
    
    CalendarMonthModel *monthModel = [calendarModels objectAtIndex:section];
    
    return monthModel.totalItemCount;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath {
    
    if (kind == UICollectionElementKindSectionHeader) {
        
        id header = [_calendarCollectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:headerIdentifier forIndexPath:indexPath];
        
        CalendarHeaderCollectionViewCell *headerCell = (CalendarHeaderCollectionViewCell*)header;
        
        NSString *month = [monthsString objectAtIndex:indexPath.section];
        
        if (month) {
//            
//            month = [month stringByReplacingOccurrencesOfString:@"月" withString:GetStringWithKey(@"月")];
            headerCell.monthLabel.text = month;
        }
        
        return (UICollectionReusableView*)headerCell;
        
    }
    
    return nil;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    CalendarCollectionViewCell *cell = [_calendarCollectionView dequeueReusableCellWithReuseIdentifier:cellIdentifier forIndexPath:indexPath];
    
    CalendarMonthModel *monthModel = [calendarModels objectAtIndex:indexPath.section];
    
    NSInteger row = indexPath.row;
    NSInteger emptyCount = monthModel.emptyCellCount;
    
    cell.dateLabel.text = @"";
    
    if (row < monthModel.emptyCellCount) {
        [cell setEmptyUI];
        [cell setUserInteractionEnabled:NO];
    }
    else {
        NSInteger day = indexPath.row - emptyCount + 1;
        
        cell.dateLabel.text = [NSString stringWithFormat:@"%ld", (long)day];
        
        [self canbeSelectCell:cell monthModel:monthModel day:day];
        NSDate *dateAtCell = [DateGenerator getDateFromInt:monthModel.year month:monthModel.month day:day];
        
        //讀取預設日期
        NSString *dateKey = self.calendarType == Start? @"start":@"end";
        
        if ([self isEqualDay:dateAtCell CompareDate: customDuration[dateKey]] &&
            selectedIndexes.count == 0 &&
            ![selectedIndexes containsObject:indexPath]) {
            [selectedIndexes removeAllObjects];
            [selectedIndexes addObject:indexPath];
        }
        
        if (cell.statusType == CalendarCollectionViewCellTypeForNormal && [selectedIndexes containsObject:indexPath]) {
            [cell setSelectedState];
        }
    }
    return cell;
}

- (BOOL)isEqualDay:(NSDate *)data CompareDate:(NSDate *)compareDate {
    if (data == nil || compareDate == nil) {
        return NO;
    }
    
    NSCalendar *cal = [NSCalendar currentCalendar];
    NSDateComponents *components = [cal components:(NSCalendarUnitEra | NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay) fromDate:data];
    NSDate *_date = [cal dateFromComponents:components];
    components = [cal components:(NSCalendarUnitEra | NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay) fromDate:compareDate];
    NSDate *_compareDate = [cal dateFromComponents:components];
    return [_date isEqualToDate:_compareDate];
}

#pragma mark - Date Method
- (void)initCalendar {
    
    monthsString = [DateGenerator getMonthsNameArray];
    
    if (!firstDate) {
        firstDate = [NSDate date];
    }
    
    NSDateComponents *components = [DateGenerator getDateComponentsFrom:firstDate];
    
    NSInteger year = components.year;
    
    [self reloadCalendarModels:year];
}

- (void)reloadCalendarModels:(NSInteger)year {
    
    NSDate *firstDate = [DateGenerator getDateFromInt:year month:1 day:1];
    
    self.yearLabel.text = [NSString stringWithFormat:@"%ld", year];
    
    [calendarModels removeAllObjects];
    
    for (int i = 0; i < 12; i++) {
        CalendarMonthModel *model = [calendarProvider getCalendarMonthModel:firstDate offset:i];
        [calendarModels addObject:model];
    }
    
    [_calendarCollectionView reloadData];
    int monthInt = [[DateGenerator getMonthString:[NSDate date]] intValue]-1;
    NSUInteger indexs2[] = {monthInt, 1};
    NSIndexPath *datePath = [NSIndexPath indexPathWithIndexes:indexs2 length:2];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.calendarCollectionView scrollToItemAtIndexPath:datePath
                                        atScrollPosition:UICollectionViewScrollPositionTop
                                                animated:YES];
    });
    
}

#pragma mark - Btn Action

- (IBAction)lastYearButtonTapped:(id)sender {
    
    CalendarMonthModel *firstModel = [calendarModels firstObject];
    
    NSInteger currentYear = firstModel.year;
    
    if (currentYear > [self minYear]) {
        currentYear -= 1;
        [self reloadCalendarModels:currentYear];
    }
    
}

- (IBAction)nextYearButtonTapped:(id)sender {
    
    CalendarMonthModel *firstModel = [calendarModels firstObject];

    NSInteger currentYear = firstModel.year;
    
//    if ([self maxYear] > currentYear) {
        currentYear += 1;
        [self reloadCalendarModels: currentYear];
//    }
    
}

- (IBAction)cancelButtonTapped:(id)sender {
    if (delegate && [delegate respondsToSelector: @selector(cancelSelectedDate)]) {
        [delegate cancelSelectedDate];
    }
    
    [self dismissViewControllerAnimated:YES completion:nil];
    
}

- (IBAction)okButtonTapped:(id)sender {
    
    // 沒選到要不要發動?
    if (delegate && selectedDate) {
        [delegate calendarDidSelect:selectedDate calendarType:calendarType];
    }
    
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: CalendarWithSwift/CalendarWithSwift/CustomCalendar/CalendarViewController.swift
//
//  CalendarViewController.swift
//  CalendarWithSwift
//
//  Created by YuChen Hsu on 2019/8/16.
//  Copyright © 2019 YuChen Hsu. All rights reserved.
//

import UIKit

enum CalendarType: Int {
    case Start = 1,
    End
}

protocol CalendarVCDelegate: AnyObject {
    
    func calendarDidSelect(_ date:Date,calendarType:CalendarType)
    
    
}

extension CalendarVCDelegate {
    
    func cancelSelectedDate() {
        
    }
}

class CalendarViewController: UIViewController,UICollectionViewDelegate,UICollectionViewDataSource,UICollectionViewDelegateFlowLayout {
    
    var headerIdentifier : String!
    var cellIdentifier : String!
    var calendarProvider : CalendarProvider!
    var monthsString = [Any]()
    var calendarModels = Array<CalendarMonthModel>()
    var selectedIndexes = Array<IndexPath>()
    var selectedDate : Date!
    var today = Date.init()
    
    @IBOutlet weak var contentView: UIView!
    @IBOutlet weak var shadowView: UIView!
    
    @IBOutlet weak var containerTopConstraint: NSLayoutConstraint!
    @IBOutlet weak var yearLabel: UILabel!
    
    @IBOutlet weak var sundayLabel: UILabel!
    @IBOutlet weak var mondayLabel: UILabel!
    @IBOutlet weak var tuesdayLabel: UILabel!
    @IBOutlet weak var wednesdayLabel: UILabel!
    @IBOutlet weak var thursdayLabel: UILabel!
    @IBOutlet weak var fridayLabel: UILabel!
    @IBOutlet weak var saturdayLabel: UILabel!
    
    @IBOutlet weak var calendarCollectionView: UICollectionView!
    
    @IBOutlet weak var cancelButton: UIButton!
    @IBOutlet weak var okButton: UIButton!

    weak var delegate : CalendarVCDelegate?
    var calendarType : CalendarType!
    var firstDate : Date!
    var customDuration = Dictionary<String, Date>()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if customDuration.count > 0 {
            // 有選擇過日期
            if customDuration.count == 2 {
                // 起迄都選過了,依照進入點
                let startDate:Date = customDuration["start"]!
                let endDate:Date = customDuration["end"]!
                
                if calendarType == CalendarType.Start{
                    firstDate = startDate
                }else{
                    firstDate = endDate
                }
            }else{
                //只選過一個
                
                let keys = Array(customDuration.keys)
                firstDate = customDuration[keys[0]]
            }
        }else{
            // 如果沒有選擇,就從今天
            firstDate = Date.init()
        }
        
        self.registerCell()
        self.setupUI()
        
        calendarProvider = CalendarProvider()
        // Do any additional setup after loading the view.
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.calendarCollectionView.reloadData()
        self.initCalendar()
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        // 防止跑版
//        if #available(iOS 11.0, *) {
//            containerTopConstraint.constant = 57.0
//        } else {
//            containerTopConstraint.constant = 77.0
//        }
        
        self.view.layoutIfNeeded()
    }
    
    // MARK: Setup UI
    func registerCell() {
        cellIdentifier = "DateCell"
        calendarCollectionView.register(UINib(nibName: "CalendarCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: cellIdentifier)
        
        headerIdentifier = "HeaderCell"
        calendarCollectionView.register(UINib(nibName: "CalendarHeaderCollectionViewCell", bundle: nil), forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: headerIdentifier)
    }
    
    func setupUI() {
        cancelButton.setTitle("Cancel", for: .normal)
        okButton.setTitle("Confirm", for: .normal)
        sundayLabel.text = "Sun"
        mondayLabel.text = "Mon"
        tuesdayLabel.text = "Tue"
        wednesdayLabel.text = "Wed"
        thursdayLabel.text = "Thu"
        fridayLabel.text = "Fri"
        saturdayLabel.text = "Sat"
        
        cancelButton.layer.borderWidth = 1.0
        cancelButton.layer.borderColor = UIColor.init(red: 0.0/255.0, green: 158.0/255.0, blue: 73.0/255.0, alpha: 1.0).cgColor
        
        shadowView.layer.shadowColor   = UIColor.black.cgColor
        shadowView.layer.shadowOffset  = CGSize.init(width: 0, height: 0)
        shadowView.layer.shadowOpacity = 0.1
        shadowView.layer.masksToBounds = false
        
        let shadowInsets = UIEdgeInsets(top: 0, left: 0, bottom: -1.5, right: 0)
        let shadowPath = UIBezierPath(rect: shadowView.bounds.inset(by: shadowInsets))
        shadowView.layer.shadowPath = shadowPath.cgPath
    }
    
    // MARK: Override method
    func maxYear() -> NSInteger {
        return NSIntegerMax
    }
    
    func minYear() -> NSInteger {
        return 1975
    }

    func canbeSelectCell(cell:CalendarCollectionViewCell, model:CalendarMonthModel, day:NSInteger) -> CalendarCollectionViewCell {
        cell.isUserInteractionEnabled = true
        cell.setNormalState()
        return cell
    }

    // MARK: UICollectionViewFlowlayoutDelegate
    func collectionView(_ collectionView: UICollectionView,
                                 layout collectionViewLayout: UICollectionViewLayout,
                                 referenceSizeForHeaderInSection section: Int) -> CGSize{
        return CGSize(width: calendarCollectionView.frame.size.width, height: 50.0)
    }

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        minimumInteritemSpacingForSectionAt section: Int) -> CGFloat{
        return 0.0
    }
    
    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        minimumLineSpacingForSectionAt section: Int) -> CGFloat{
        return 0.0
    }
    
    func collectionView(_ collectionView: UICollectionView,
    layout collectionViewLayout: UICollectionViewLayout,
    sizeForItemAt indexPath: IndexPath) -> CGSize{
        let itemWidth:CGFloat = calendarCollectionView.frame.size.width / 7.0
        
        let itemHeight:CGFloat = itemWidth
        
        return CGSize(width: itemWidth, height: itemHeight)
    }
    
    // MARK: CollectionView Delegate
    func collectionView(_ collectionView: UICollectionView,
                        didSelectItemAt indexPath: IndexPath){
        
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int{
        return calendarModels.count
    }
    
    func collectionView(_ collectionView: UICollectionView,
                        numberOfItemsInSection section: Int) -> Int{
        
        if calendarModels.count == 0 {
            return 0
        }
        
        let monthModel:CalendarMonthModel = calendarModels[section] 
        
        return monthModel.totalItemCount
    }
    
     func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        
        if kind == UICollectionView.elementKindSectionHeader {
            let header:Any = calendarCollectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: headerIdentifier, for: indexPath)
            
            let headerCell:CalendarHeaderCollectionViewCell = header as! CalendarHeaderCollectionViewCell
            
            let month:String = monthsString[indexPath.section] as! String
            
            headerCell.monthLabel.text = month
            
            return headerCell as UICollectionReusableView;
            
        }
        
        return UICollectionReusableView.init()
    }
    
    func collectionView(_ collectionView: UICollectionView,
                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell{
        
        var cell:CalendarCollectionViewCell = calendarCollectionView.dequeueReusableCell(withReuseIdentifier: cellIdentifier, for: indexPath) as! CalendarCollectionViewCell
        
        let monthModel:CalendarMonthModel = calendarModels[indexPath.section]
        
        let row:NSInteger = indexPath.row
        
        let emptyCount:NSInteger = monthModel.emptyCellCount
        
        cell.dateLabel.text = ""
        
        if row < monthModel.emptyCellCount {
            cell.setEmptyUI()
            cell.isUserInteractionEnabled = false
        }else{
            let day:NSInteger = indexPath.row - emptyCount + 1
            
            cell.dateLabel.text = String(format:"%ld", day)
            
            cell = self.canbeSelectCell(cell: cell, model: monthModel, day: day)
            
            let dateAtCell:Date = DateGenerator.getDateFrom(year: monthModel.year, month: monthModel.month, day: day)
            
            var dateKey:String = "end"
            
            if self.calendarType == .Start{
                dateKey = "start"
            }
            
            if self.isEqualDay(date: dateAtCell, compareDate: customDuration[dateKey] as? Date) && selectedIndexes.count == 0 && !selectedIndexes.contains(indexPath){
                selectedIndexes.removeAll(keepingCapacity: false)
                selectedIndexes.append(indexPath)
            }
            
            if cell.statusType == .calendarCollectionViewCellTypeForNormal && selectedIndexes.contains(indexPath){
                cell.setSelectedState()
            }
        }
        return cell
    }
    
    func isEqualDay(date:Date?, compareDate:Date?) -> Bool {
        if date == nil || compareDate == nil{
            return false
        }
        
        let cal:Calendar = Calendar.current
        
        var components:DateComponents = cal.dateComponents([.era,.year,.month, .day], from: date! )
        
        let newDate:NSDate = cal.date(from: components)! as NSDate
        
        components = cal.dateComponents([.era,.year,.month, .day], from: compareDate! )
        
        let newCompareDate:Date = cal.date(from: components)!
        
        return newDate.isEqual(to: newCompareDate)
    }
    
    //MARK: Date Method
    func initCalendar(){
        
        monthsString = DateGenerator.getMonthsNameArray()
        
        let components:DateComponents = DateGenerator.getDateComponentsFrom(date: firstDate)
        
        let year:NSInteger = components.year!
        
        self.reloadCalendarModels(year: year)
    }
    
    func reloadCalendarModels(year:NSInteger){
        
        let firstDate:Date = DateGenerator.getDateFrom(year: year, month: 1, day: 1)
        
        self.yearLabel.text = String(format:"%ld", year)
        
        calendarModels.removeAll(keepingCapacity: false)
        
        for i in (0...11){
            let model:CalendarMonthModel = calendarProvider.getCalendarMonthModel(date: firstDate, offset: i)
            calendarModels.append(model)
        }
        
        calendarCollectionView.reloadData()
        let monthInt:Int = Int(DateGenerator.getMonthString(date: Date.init()))! - 1
        
        let indexs2:[Int] = [monthInt, 1]
        let datePath:IndexPath = (IndexPath.init(indexes: indexs2))
        
        DispatchQueue.main.async {
            self.calendarCollectionView.scrollToItem(at: datePath, at: .top, animated: true)
        }
    }
    
    // MARK: Btn Action
    @IBAction func lastYearButtonTapped(_ sender: Any) {
        
        let firstModel:CalendarMonthModel = calendarModels.first as! CalendarMonthModel
        
        var currentYear:NSInteger = firstModel.year
        
        if currentYear > self.minYear() {
            currentYear = currentYear - 1
            self.reloadCalendarModels(year: currentYear)
        }
    }
    
    @IBAction func nextYearButtonTapped(_ sender: Any) {
        
        let firstModel:CalendarMonthModel = calendarModels.first as!CalendarMonthModel
        
        var currentYear:NSInteger = firstModel.year
            currentYear = currentYear + 1
            self.reloadCalendarModels(year: currentYear)
    }
    
    @IBAction func cancelButtonTapped(_ sender: Any) {
        
        let delegateMethod = delegate?.cancelSelectedDate()
        
        if (delegate != nil), (delegateMethod != nil) {
            delegate!.cancelSelectedDate()
        }
        
        self.dismiss(animated: true, completion: nil)
        
        }
        
    @IBAction func okButtonTapped(_ sender: Any) {
        // 沒選到要不要發動?
        let delegateMethod = delegate?.calendarDidSelect(selectedDate, calendarType: calendarType)
        
        if (delegate != nil), (delegateMethod != nil) {
            delegate!.calendarDidSelect(selectedDate, calendarType: calendarType)
        }
        
        self.dismiss(animated: true, completion: nil)
        }
}
